homeassistant/components/tuya/__init__.py

```diff
@@ -52,17 +56,41 @@
 )
 
 
-def setup(hass, config):
+def setup(hass, config, retry_delay=FIRST_RETRY_TIME):
     """Set up Tuya Component."""
 
+    _LOGGER.debug("Setting up integration")
+
     tuya = TuyaApi()
     username = config[DOMAIN][CONF_USERNAME]
     password = config[DOMAIN][CONF_PASSWORD]
     country_code = config[DOMAIN][CONF_COUNTRYCODE]
     platform = config[DOMAIN][CONF_PLATFORM]
 
+    try:
+        tuya.init(username, password, country_code, platform)
+    except (TuyaNetException, TuyaServerException):
+
+        _LOGGER.warning(
+            "Connection error during integration setup. Will retry in %s seconds",
+            retry_delay,
+        )
+
+        def retry_setup(now):
+            """Retry setup if a error happens on tuya API."""
+            setup(hass, config, retry_delay=min(2 * retry_delay, MAX_RETRY_TIME))
+
+        call_later(hass, retry_delay, retry_setup)
+
+        return True
+
+    except TuyaAPIException as ex:
+        _LOGGER.error(
+            "Connection error during integration setup. Error: %s", str(ex),
```

Comment:
Do we know that the connection error is due to bad authentication here? If so we should be more specific in the error message.

We don't need to copy the exception argument to string.
```suggestion
            "Connection error during integration setup. Error: %s", exc,
```

Reply:
Detail of the error is in the exception text

Reply:
Ok. And we know that these errors are not recoverable, in contrast to the errors above?

Reply:
Yes, because recoverable error was raised before with other exceptions.